File: DP/src/main/java/complete_backpack/solution518.java
package complete_backpack;

public class solution518 {
    public int change(int amount, int[] coins) {
        int n = coins.length;
        int[][] opt = new int[n + 1][amount + 1];

        opt[0][0] = 1;

        for (int i = 1; i < n + 1; i++) {
            int val = coins[i - 1];

            for (int j = 0; j < amount + 1; j++) {
                opt[i][j] = opt[i - 1][j];

                for (int k = 1; k * val <= j; j++) {
                    opt[i][j] = opt[i - 1][j - k * val] + opt[i][j];
                }
            }

        }

        return opt[n][amount];

    }

}

File: DP/src/main/python/完全背包/solution518.py
from typing import List


class Solution:
    def change(self, amount: int, coins: List[int]) -> int:

        n = len(coins)  # N种物品

        opt = [[0 for _ in range(amount + 1)] for _ in range(n + 1)]

        opt[0][0] = 1

        for i in range(1, n + 1):
            value = coins[i - 1]
            for j in range(amount + 1):
                opt[i][j] = opt[i - 1][j]

                k = 1
                while k * value <= j:
                    opt[i][j] = opt[i][j] + opt[i - 1][j - k * value]
                    k += 1

        return opt[n][amount]


s = Solution()

amount = 5
coins = [1, 2, 5]

res = s.change(amount, coins)
print(res)
